modules/vault/src/main/java/org/apache/ignite/internal/vault/impl/VaultServiceImpl.java

```
@@ -0,0 +1,130 @@
+        synchronized (mux) {
+            return CompletableFuture.completedFuture(IgniteUtils.bytesToLong(storage.get(APPLIED_REV), 0));
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public CompletableFuture<Void> put(ByteArray key, byte[] val) {
+        synchronized (mux) {
+            storage.put(key, val);
+
+            watcher.notify(new VaultEntry(key, val));
+
+            return CompletableFuture.allOf();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public CompletableFuture<Void> remove(ByteArray key) {
+        synchronized (mux) {
+            storage.remove(key);
+
+            return CompletableFuture.allOf();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public Iterator<VaultEntry> range(ByteArray fromKey, ByteArray toKey) {
+        synchronized (mux) {
+            return new ArrayList<>(storage.subMap(fromKey, toKey).entrySet())
+                .stream()
+                .map(e -> new VaultEntry(e.getKey(), e.getValue()))
+                .iterator();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public @NotNull CompletableFuture<IgniteUuid> watch(@NotNull VaultWatch vaultWatch) {
+        synchronized (mux) {
+            return watcher.register(vaultWatch);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public @NotNull CompletableFuture<Void> stopWatch(@NotNull IgniteUuid id) {
+        synchronized (mux) {
+            watcher.cancel(id);
+
+            return CompletableFuture.allOf();
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override public @NotNull CompletableFuture<Void> putAll(@NotNull Map<ByteArray, byte[]> vals, long revision) {
+        synchronized (mux) {
+            byte[] appliedRevBytes = storage.get(APPLIED_REV);
+
+            long appliedRevision = appliedRevBytes != null ? IgniteUtils.bytesToLong(appliedRevBytes, 0) : 0L;
+
+            if (revision < appliedRevision)
```

Instead of silently ignoring `revision < appliedRevision` I'd rather return false in order to notify user about safe but unexpected behavior.

---

There are no any revision in vault. It is mixing of business logic and storage functionality. Must be removed to the manager.

---

We agreed with @sanpwc to throw `CheckedException` when `revision < appliedRevision`. Also agreed, that this should be moved to manger